README.md
## Sonic Pi

This is the source for the Sonic Pi music programming environment. The dependencies for this are:

* supercollider
* ruby1.9.1
* libqscintilla2-8

Compilation
-----------

Build dependencies:

* g++
* make
* libqscintilla2
* qt

To install dependencies for Debian-based systems:
```
sudo apt-get install g++ make libqt4-dev libqscintilla2-dev
```

Or for Red Hat-based:
```
sudo yum install g++ make qt-devel qscintilla-devel
```

To compile:
```
qmake ./app/gui/make-music.pro
make
```

Running
-------

Initially run with
```
./bin/make-music
```
to create directories, setup the workspaces and position the window.

However, if the window positioning is not required, then the program can subsequently be run directly with

```
./make-music
```

File System Structure
---------------------

`./app/gui/` contains the main QT program source files.

`./app/scripts/` contains the scripts to be called when a song is played. They define things such as the `play`, `play_chord` and `play_pattern` functions.

Notes
-----

The current implementation assumes the execution context is a Raspberry Pi. Patches for other platforms will be happily considered.


debian/changelog
make-music (3.16.0-0) unstable; urgency=low

  * Clean up fully when updating

 -- Team Kano <dev@kano.me>  Tue, 25 Sep 2018 16:01:00 +0100

make-music (3.14.0-0) unstable; urgency=low

  * Applied changes from v3.9.2 es_AR backports to latest

 -- Team Kano <dev@kano.me>  Mon, 16 Oct 2017 15:12:00 +0100

make-music (3.9.2-0) unstable; urgency=low

  * Encoded es_AR challenges to Latin1 due to Sonic Pi bug (backport)
  * Fixed loading files with UTF-8 characters

 -- Team Kano <dev@kano.me>  Wed, 11 Oct 2017 18:31:00 +0100

make-music (3.9.1-0) unstable; urgency=low

  * Fixed missing comment on es_AR challenge two (backport)

 -- Team Kano <dev@kano.me>  Fri, 21 Jul 2017 13:12:47 +0100

make-music (2.10-0) unstable; urgency=low

  * Updated version to Sonic-Pi 2.10.0 (Cowbell)

 -- Team Kano <dev@kano.me>  Tue, 07 Jun 2016 14:00:00 +0000

make-music (2.9-0) unstable; urgency=low

  * Updated version to Sonic-Pi 2.9.0 (Venster)

 -- Team Kano <dev@kano.me>  Mon, 18 Jan 2016 15:15:00 +0000

make-music (2.7-0) unstable; urgency=low

  * Updated version to Sonic-Pi 2.7.0 (Rerezzed)

 -- Team Kano <dev@kano.me>  Tue, 22 Sep 2015 15:00:00 +0000

make-music (2.5-0) unstable; urgency=low

  * Updated version to Sonic-Pi 2.5.0 (Craft)

 -- Team Kano <dev@kano.me>  Wed, 18 Feb 2015 17:02:39 +0000

make-music (2.4-0) unstable; urgency=low

  * Updated version to Sonic-Pi 2.4.0

 -- Team Kano <dev@kano.me>  Wed, 18 Feb 2015 17:02:39 +0000

make-music (2.1-1) unstable; urgency=low

  * Updated version to Sonic-Pi 2.1.1

 -- Team Kano <dev@kano.me>  Tue, 16 Dec 2014 11:10:39 +0000

make-music (2.0-1) unstable; urgency=low

  * Updated version to Sonic-Pi 2

 -- Team Kano <dev@kano.me>  Mon, 22 Sep 2014 18:10:39 +0000

make-music (1.1-1) unstable; urgency=low

  * Reviewing the package for release

 -- Team Kano <dev@kano.me>  Tue, 17 Mar 2014 14:10:39 +0000

make-music (1.0-32) unstable; urgency=low

  * Package rebuilt, updated to revision f41b602e.

 -- Team Kano <dev@kano.me>  Mon, 03 Mar 2014 21:06:15 +0000

make-music (1.0-31) unstable; urgency=low

  * Package rebuilt, updated to revision 13518d8a.

 -- Team Kano <dev@kano.me>  Mon, 24 Feb 2014 14:36:47 +0000

make-music (1.0-30) unstable; urgency=low

  * Package rebuilt, updated to revision 6130329e.

 -- Team Kano <dev@kano.me>  Wed, 05 Feb 2014 02:21:46 +0000

make-music (1.0-29) unstable; urgency=low

  * Package rebuilt, updated to revision 820abf62.

 -- Team Kano <dev@kano.me>  Thu, 30 Jan 2014 02:19:27 +0000

make-music (1.0-28) unstable; urgency=low

  * Package rebuilt, updated to revision 60118828.

 -- Team Kano <dev@kano.me>  Fri, 17 Jan 2014 09:29:38 +0000

make-music (1.0-27) unstable; urgency=low

  * Package rebuilt, updated to revision 39798799.

 -- Team Kano <dev@kano.me>  Fri, 17 Jan 2014 02:24:00 +0000

make-music (1.0-26) unstable; urgency=low

  * Package rebuilt, updated to revision f908f809.

 -- Team Kano <dev@kano.me>  Mon, 13 Jan 2014 11:42:45 +0000

make-music (1.0-25) unstable; urgency=low

  * Package rebuilt, updated to revision 7edbb59c.

 -- Team Kano <dev@kano.me>  Sat, 11 Jan 2014 02:43:29 +0000

make-music (1.0-24) unstable; urgency=low

  * Package rebuilt, updated to revision 6e69b353.

 -- Team Kano <dev@kano.me>  Thu, 02 Jan 2014 03:04:37 +0000

make-music (1.0-23) unstable; urgency=low

  * Package rebuilt, updated to revision 27a4499a.

 -- Team Kano <dev@kano.me>  Wed, 01 Jan 2014 03:04:00 +0000

make-music (1.0-22) unstable; urgency=low

  * Package rebuilt, updated to revision e1293de3.

 -- Team Kano <dev@kano.me>  Tue, 31 Dec 2013 03:04:46 +0000

make-music (1.0-21) unstable; urgency=low

  * Package rebuilt, updated to revision 017dcea6.

 -- Team Kano <dev@kano.me>  Mon, 30 Dec 2013 03:03:26 +0000

make-music (1.0-20) unstable; urgency=low

  * Package rebuilt, updated to revision 66bb64c2.

 -- Team Kano <dev@kano.me>  Sun, 29 Dec 2013 03:03:20 +0000

make-music (1.0-19) unstable; urgency=low

  * Package rebuilt, updated to revision 9e7b508b.

 -- Team Kano <dev@kano.me>  Sat, 28 Dec 2013 03:03:20 +0000

make-music (1.0-18) unstable; urgency=low

  * Package rebuilt, updated to revision 08d93946.

 -- Team Kano <dev@kano.me>  Fri, 27 Dec 2013 03:03:40 +0000

make-music (1.0-17) unstable; urgency=low

  * Package rebuilt, updated to revision 42df1ff9.

 -- Team Kano <dev@kano.me>  Thu, 26 Dec 2013 03:03:04 +0000

make-music (1.0-16) unstable; urgency=low

  * Package rebuilt, updated to revision 907ddf88.

 -- Team Kano <dev@kano.me>  Wed, 25 Dec 2013 03:03:28 +0000

make-music (1.0-15) unstable; urgency=low

  * Package rebuilt, updated to revision 705e40f4.

 -- Team Kano <dev@kano.me>  Tue, 24 Dec 2013 03:03:24 +0000

make-music (1.0-14) unstable; urgency=low

  * Package rebuilt, updated to revision fd6686f3.

 -- Team Kano <dev@kano.me>  Mon, 23 Dec 2013 03:03:25 +0000

make-music (1.0-13) unstable; urgency=low

  * Package rebuilt, updated to revision d3de504d.

 -- Team Kano <dev@kano.me>  Sun, 22 Dec 2013 03:03:31 +0000

make-music (1.0-12) unstable; urgency=low

  * Package rebuilt, updated to revision 52d5d36d.

 -- Team Kano <dev@kano.me>  Sat, 21 Dec 2013 03:03:30 +0000

make-music (1.0-11) unstable; urgency=low

  * Package rebuilt, updated to revision 6f103711.

 -- Team Kano <hello@kano.me>  Fri, 20 Dec 2013 14:06:34 +0000

make-music (1.0-10) unstable; urgency=low

  * Package rebuilt, updated to revision 40bb7905.

 -- Team Kano <hello@kano.me>  Thu, 19 Dec 2013 02:26:35 +0000

make-music (1.0-9.1) unstable; urgency=low

  * Non-maintainer upload.
  * Package rebuilt, updated to revision ba320742.

 -- Team Kano <hello@kano.me>  Tue, 17 Dec 2013 02:51:43 +0000

make-music (1.0-9) unstable; urgency=low

  * Package rebuilt, updated to revision 5316094c.

 -- Team Kano <dev@kano.me>  Fri, 13 Dec 2013 16:43:37 +0000

make-music (1.0-8) unstable; urgency=low

  * Package rebuilt, updated to revision 38242a4b.

 -- Team Kano <dev@kano.me>  Thu, 12 Dec 2013 16:16:31 +0000


app/gui/lexer.h
/*
# lexer.h
#
# Copyright (C) 2013 Kano Computing Ltd.
# License: http://opensource.org/licenses/MIT
#
# This extends the QsciLexerRuby class from Scintilla for the purpose
# of formatting code specific to Sonic Pi
*/

#ifndef LEXER_H
#define LEXER_H

#ifdef __APPLE__
extern "C++" {
#endif

#include <qobject.h>

#include <Qsci/qsciglobal.h>
//#include <Qsci/qscilexer.h>

#include <Qsci/qscilexerruby.h>

//! \brief The Lexer class encapsulates the Sonic Pi lexer.
class QSCINTILLA_EXPORT Lexer : public QsciLexerRuby
{
    Q_OBJECT

public:
    //! This enum defines the meanings of the different styles used by the
    //! Ruby lexer.
    enum {
        //! The default.
        Default = 0,

        //! An error.
        Error = 1,

        //! A comment.
        Comment = 2,

        //! A POD.
        POD = 3,

        //! A number.
        Number = 4,

        //! A keyword.
        Keyword = 5,

        //! A double-quoted string.
        DoubleQuotedString = 6,

        //! A single-quoted string.
        SingleQuotedString = 7,

        //! The name of a class.
        ClassName = 8,

        //! The name of a function or method.
        FunctionMethodName = 9,

        //! An operator.
        Operator = 10,

        //! An identifier
        Identifier = 11,

        //! A regular expression.
        Regex = 12,

        //! A global.
        Global = 13,

        //! A symbol.
        Symbol = 14,

        //! The name of a module.
        ModuleName = 15,

        //! An instance variable.
        InstanceVariable = 16,

        //! A class variable.
        ClassVariable = 17,

        //! Backticks.
        Backticks = 18,

        //! A data section.
        DataSection = 19,

        //! A here document delimiter.
        HereDocumentDelimiter = 20,

        //! A here document.
        HereDocument = 21,

        //! A %q string.
        PercentStringq = 24,

        //! A %Q string.
        PercentStringQ = 25,

        //! A %x string.
        PercentStringx = 26,

        //! A %r string.
        PercentStringr = 27,

        //! A %w string.
        PercentStringw = 28,

        //! A demoted keyword.
        DemotedKeyword = 29,

        //! stdin.
        Stdin = 30,

        //! stdout.
        Stdout = 31,

        //! stderr.
        Stderr = 40
    };

    //! Construct a Lexer with parent \a parent.  \a parent is
    //! typically the QsciScintilla instance.
    Lexer(QObject *parent = 0 );

    //! Destroys the Lexer instance.
    virtual ~Lexer();

    //! Returns the name of the lexer.  Some lexers support a number of
    //! languages.
    const char *lexer() const;

    //! \internal Returns a space separated list of words or characters in
    //! a particular style that define the end of a block for
    //! auto-indentation.  The style is returned via \a style.
    const char *blockEnd(int *style = 0) const;

    //! \internal Returns a space separated list of words or characters in
    //! a particular style that define the start of a block for
    //! auto-indentation.  The styles is returned via \a style.
    const char *blockStart(int *style = 0) const;

    //! \internal Returns a space separated list of keywords in a
    //! particular style that define the start of a block for
    //! auto-indentation.  The style is returned via \a style.
    const char *blockStartKeyword(int *style = 0) const;

    //! \internal Returns the style used for braces for brace matching.
    int braceStyle() const;

    //! Returns the foreground colour of the text for style number \a style.
    //!
    //! \sa defaultpaper()
    QColor defaultColor(int style) const;

    //! Returns the end-of-line fill for style number \a style.
    bool defaultEolFill(int style) const;

    //! Returns the font for style number \a style.
    QFont defaultFont(int style) const;

    //! Returns the background colour of the text for style number \a style.
    //!
    //! \sa defaultColor()
    QColor defaultPaper(int style) const;

    //! Returns the set of keywords for the keyword set \a set recognised
    //! by the lexer as a space separated string.
    const char *keywords(int set) const;

    //! Returns the descriptive name for style number \a style.  If the style
    //! is invalid for this language then an empty QString is returned.  This
    //! is intended to be used in user preference dialogs.
    QString description(int style) const;

};

#ifdef __APPLE__
}
#endif

#endif
